### PR TITLE
Java MAVLink v2

### PR DESCRIPTION
**Pull request by len0rd**
This replaces #341 (I could try to PR this into #341, but I'm unsure magicrub is interested in getting involved again). 341 replaced #323 which was contributed to in #190, #320 and #317. Which build off #290 which builds off #127

A lot of the feature-creep changes (#190 - package renaming, json serialization) have been removed. This should be the bare minimum to get MAVLink V2 working on Java while maintaining backwards compatibility. I have kept the older commits to preserve the contributions made by everyone.

I have a few feature branches (formatting improvements, more extensive Travis CI tests) that build off of this. So far I've only verified the generated code still compiles, but I should have some more extensive unit tests finished in the next couple days.

Credited authors:
@HefnySco @Hunter522 @rrr6399 @kellyschrock @magicrub

### CODE REVIEW

File: generator/mavgen_java.py [hunk 1]
```diff
@@ -60,48 +58,57 @@ def generate_CRC(directory, xml):
...
-    public void finish_checksum(int msgid) {
-        update_checksum(MAVLINK_MESSAGE_CRCS[msgid]);
+     * Finish the CRC calculation of a message, by running the CRC with the
+     * Magic Byte.
+     *
+     * @param msgid The message id number
+     * @return boolean True if the checksum was successfully finished. Otherwise false
+     */
+    public boolean finish_checksum(int msgid) {
+        if (MAVLINK_MESSAGE_CRCS.containsKey(msgid)) {
```
Comment (len0rd) [inline generator/mavgen_java.py:102]: @WickedShell My solution to our [discussion](https://github.com/ArduPilot/pymavlink/pull/341#discussion_r305652662) in the previous PR is to return a boolean back up to the Parser. If CRC extra does not exist, a false is passed back up and the current packet is tossed out

**Comment by HefnySco** [comment]
I generated code using this PR and tried it on my application and again it worked great.
It can detect mavlink v1 & v2.
Thank you.

**Comment by davidbuzz** [comment]
i support this going in, fwiw. :-)

**Comment by len0rd** [comment]
Rebased to resolve conflict with new doc strings in mavgen_java.py

**Comment by magicrub** [comment]
Thanks @len0rd for owning this. Maybe as an Xmas present a maintainer will merge this or at least tell us what's blocking it. It's quite embarrassing that it's behind held up this long.

**Comment by tridge** [comment]
I'm going to merge this as it is being actively used, and is clearly better than what we have now

**Comment by davidbuzz** [comment]
@len0rd  - would you be interested in becoming the maintainer for the Java mavlink bindings?   Tridge mentioned in discord that we're still looking for an ongoing maintainer.... or perhaps u know someone who might ?   no-guarantees of course, as its up to tridge but i though it worth  asking.

**Comment by rrr6399** [comment]
Thanks for finishing this up!

I was curious why the Java package names were changed to MAVLink from mavlink?

Merry Christmas!

**Comment by len0rd** [comment]
@rrr6399 The original v1 code used `MAVLink` (not `mavlink`), so the maintainers requested it remain the same for the sake of backwards compatibility. If you can think of a clever way to change the package to `mavlink` (which imo is what it should be) while maintaining backwards compatibility I'm all for it.

At a higher-level, these are the types of things that would benefit from pymavlink having versioned releases - allowing us to have a major release with a breaking change like this.

Merry Christmas!

**Comment by len0rd** [comment]
@davidbuzz yea, I would be happy to help out! Though, there are contributors to this PR that are likely much more qualified than me. I mainly use the C bindings, and only occasionally mess with the java side.

Regardless, I'm planning to submit another PR with some automated CI testing that should ease java binding verification.

**Comment by davidbuzz** [comment]
>another PR with some automated CI testing - thanks... i'm not sure whether u noticed, but I've been doing similar with the JavaScript/Node binding recently...  (automated test suit comparing their input/parsing and output assembly against the C bindings , in my case )

**Comment by rrr6399** [comment]
> @rrr6399 The original v1 code used `MAVLink` (not `mavlink`), so the maintainers requested it remain the same for the sake of backwards compatibility. If you can think of a clever way to change the package to `mavlink` (which imo is what it should be) while maintaining backwards compatibility I'm all for it.

I'll check out your code and see if I can add a command line flag to change the package name to mavlink with the default being MAVLink.

**Comment by davidbuzz** [comment]
@lenord  - are u on ardupilot discord? rather than abuse this close PR for chat,  we could continue this chat there if u are... https://ardupilot.org/discord ?

**Pull request merged**